File: jetbrains/src/main/kotlin/deplens/utils/RequestManager.kt
package deplens.utils

import java.util.concurrent.ConcurrentHashMap

private val maxRetryCount = 3

private val failedExpireTime = 30 * 60 * 1000; // 30 分钟

class RequestManager {

    private val failed = ConcurrentHashMap<String, List<Long>>()

    fun shouldRequest(key: String): Boolean {

        val now = System.currentTimeMillis()

        val items = failed[key]
            ?.filter { expireTime ->
                expireTime > now
            }?.let { times ->
                failed[key] = times
                times
            }

        val cnt = items?.size ?: 0

        return cnt < maxRetryCount
    }

    fun updateFailed(key: String) {
        val now = System.currentTimeMillis()
        val expireTimes = (failed[key] ?: emptyList()) + (now + failedExpireTime)
        failed[key] = expireTimes
    }

    fun getFailureCount(key: String): Int {
        val now = System.currentTimeMillis()
        val items = failed[key]?.filter { expireTime -> expireTime > now } ?: return 0
        return items.size
    }

    fun hasFailure(key: String): Boolean = getFailureCount(key) > 0

    fun clearFailure(key: String) {
        failed.remove(key)
    }
}

File: vscode/src/utils/RequestManager.ts
const MAX_RETRY_COUNT = 3;
const FAILED_EXPIRE_TIME = 30 * 60 * 1000; // 30 minutes

export class RequestManager {
  private failed = new Map<string, number[]>();

  shouldRequest(key: string): boolean {
    const now = Date.now();
    let items = this.failed.get(key) || [];

    // Filter expired failures
    items = items.filter((expireTime) => expireTime > now);
    this.failed.set(key, items);

    const cnt = items.length;
    return cnt < MAX_RETRY_COUNT;
  }

  updateFailed(key: string): void {
    const now = Date.now();
    const items = this.failed.get(key) || [];
    const expireTimes = [...items, now + FAILED_EXPIRE_TIME];
    this.failed.set(key, expireTimes);
  }
}
